compiler/rustc_serialize/src/serialize.rs

```diff
@@ -50,9 +61,12 @@ pub trait Encoder {
     // optimization that would otherwise be necessary here, likely due to the
     // multiple levels of inlining and const-prop that are needed.
     #[inline]
-    fn emit_fieldless_enum_variant<const ID: usize>(&mut self) -> Result<(), Self::Error> {
+    fn emit_fieldless_enum_variant<const ID: usize>(&mut self) {
         self.emit_usize(ID)
     }
+
+    // Consume the encoder, getting the result.
+    fn finish(self) -> Result<Self::Ok, Self::Err>;
```

Comment:
Does this need to be a method on the `Encoder` trait, or can it be a method on individual encoders?

Reply:
It can be a method on individual encoders, and I have made this change.

This did require that `CacheEncoder` be made non-generic, because it needs to call `finish` on the underlying encoder, and if `finish` isn't within the trait then a concrete type is required. This is fine in practice because it always uses a `FileEncoder`.

Other advantages of the change:
- Avoids some `unwrap` calls on `Encoder::finish`.
- Avoids some imports.
- Removes the `Ok` and `Err` associated types from trait `Encoder`.
- Removes need for the `OpaqueEncoder` trait.